### ISSUE TITLE
RyuJIT generates incorrect exception handling scopes for IL generated by Roslyn

### ISSUE THREAD

**Issue by jkotas** [dotnet/coreclr#1421]
Repro:

1. Compile https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/Regression/CLR-x86-JIT/V1-M11-Beta1/b43313/b43313.cs using Roslyn, with optimizations on.
2. Run it on CoreCLR x64

**Comment by jkotas** [comment]
Rollback https://github.com/dotnet/coreclr/commit/69dcfe966ba7bf7ffd3dd2b000943590279da358 as part of this fix

**Comment by mikedn** [comment]
A simpler and more realistic repro:

```C#
using System;

class Program
{
    static void Main()
    {
        var a = new int[7];

        try
        {
            Console.WriteLine("try");
        }
        catch (IndexOutOfRangeException)
        {
            Console.WriteLine("noway");
        }

        while (a[42] != 0)
            ;
    }
}
```

This prints "try" "noway" and crashes. Of course, "noway" should have not been printed.

Unlike the original repro this doesn't have an empty try block. That empty try block is the reason the test works fine when the code is compiled with the old C# compiler but fails when compiled with Roslyn. The old C# compiler removes that try block and the subsequent catch block and thus avoids a JIT bug. But the JIT bug exists even if the try block isn't empty.

The problem is that `fgOptWhileLoop` duplicates the while loop test in an attempt to eliminate an unconditional jump from the try block to the following block. Obviously that's not correct as it causes the loop test to be performed inside the try block thus diverting whatever exceptions it throws to the catch handler.

It's possible that this happens because `impImportLeave` did not set the `BBF_KEEP_BBJ_ALWAYS` flags when changing the jump kind from `CEE_LEAVE` to `BBJ_ALWAYS`.

It's not clear to me why an unconditional jump exists there in the first place, the while block directly follows the try block so there's no need for any jump.

**Comment by mikedn** [comment]
Note that in a future release Roslyn will get rid of empty try blocks - [#2443](https://github.com/dotnet/roslyn/issues/2443) - this test will then fail again to find JIT bugs.

**Comment by BruceForstall** [comment]
The code fix went in with 767e9175ea115425b814432c4362afb0d83b3668, mirrored from the TFS side (for various reasons). Jan has a PR going to revert the test change.
